Added support for multiple in VocabField

bworrell opened the pull request:
Added support for `mutliple` in `VocabField`. Fixes #244.

ikiril01 commented:
@bworrell awesome! All issues related to #244 appear to be fixed :+1: I'll merge this in unless @gtback objects.

gtback commented on cybox/utils/__init__.py:
```diff
@@ -73,6 +74,17 @@ def _import_submodules(pkg):
         importlib.import_module(mod_name)
 
 
+def is_sequence(item):
```
Is there a reason we want to support tuples? Why not just check if `value` is a list?

bworrell replied:
I changed all of python-stix to use the `is_sequence` method because I didn't really like the restrictive nature of requiring a `list` for input. Users may want to pass in any (non-`Mapping`) iterable object such as `set`, `tuple`, or a generator object to extract values from. I guess I didn't see any reason to raise errors when instances of those objects are passed in.

gtback replied:
I understand what you're trying to do, it just feels like an artificial distinction (you can pass in any iterable except a dictionary or string). Since we are converting it to a list anyway, I would be fine forcing people to wrap any generators in a `list()` call. On the other hand, Python's default behavior when iterating over dictionaries is to give you the keys; if someone passes in a dictionary expecting that behavior, I don't know if we should restrict it. IMO, the only thing we really want to check is that it's not a string; if we iterate over that to get characters, we'll have a problem.

That said, I'm fine leaving this; I just saw it and thought "this shouldn't be necessary", but  a quick search didn't find anything better.

bworrell replied:
I'd be totally fine with removing the `Mapping` clause.

Originally, I felt like the behavior when passing in a dict would be kinda weird (though totally consistent with other builtin methods that take iterables as parameters) so I restricted it. The more I think about it, though more I think removing that restriction _might_ be the best option, since it's pretty consistent with other Python functions.

I think I originally wrote this for python-stix because I use tuples and sets a lot and got frustrated by the `list` requirement after seeing one-too-many stack traces :)

gtback commented on cybox/common/vocabs.py:
```diff
@@ -33,21 +33,41 @@ def __init__(self, *args, **kwargs):
+        error = error_fmt % (self.name, self.type_, type(value))
+        raise ValueError(error)
+
+    def __set__(self, instance, value):
```
Any chance we can try to consolidate this behavior with `TypedField.__set__`, so that the only difference between `TypedField` and `VocabField` is the implementation of `_handle_value`?

bworrell replied:
Sure! Well, probably anyway, heh. I had thought about tweaking `TypedField`, but I wasn't sure I had implemented `multiple` correctly :)

bworrell commented:
@gtback, I took a stab at your fixes:
* Modified `TypedField` to have a `_handle_value()` method and support type inspection and casting when `multiple` is `True`. 
* Removed the `Mapping` restriction on `is_sequence()` so dictionaries return `True`.

One thing I did that I wasn't sure about was "skipping" `None` values in `__set__()` when multiple is `True`. So calling `Foo.bar = [1,2,None,3]` would turn into `[1,2,3]` internally.

I'm not sure if that seems fine **or** if it'd be better to allow `None` to persist through `__set__()`  and just have `Entity` know to skip `None` in `to_dict()` and `to_obj()` when iterating over the values.

I guess it comes down to what `None` means in a list of entities. Is there significance to that? I'm not sure there is TBH.

bworrell commented:
I thought a bit more about this last night and I think it's fine to skip `None` values in the `__set__` method (when `multiple` is `True`) since we don't allow `xsi:nil="true"` on any elements. If we allowed `nil`, it'd probably make sense to retain the `None` values.

gtback commented:
:+1: I really like how `TypedField` and `VocabField` interact with `__set__`, `_handle_value`, and the implementation of `multiple`.

I think skipping `None`s is perfectly fine.
